### PR TITLE
Convert river and tidal modules to xarray

### PR DESCRIPTION
**Pull request by akeeste**
This PR converts the river and tidal modules to be based in xarray

TODO list
- [x] for functions where pandas is an input, add allow xarray as an allowed input
- [x] for functions that take pandas or xarray input, convert the internal functionality to be xarray-based
- [x] add tests for xarray input
- [x] for functions that output pandas, add a flag to output xarray
- [x] all tests passing
- [x] river, tidal, tidal performance notebooks running

Modules completed:
- [x] river.graphics
- [x] river.performance
- [x] river.resource
- [ ] river.io.usgs
- [ ] river.io.d3d / tidal.io.d3d
- [x] tidal.graphics
- [x] tidal.performance
- [x] tidal.resource
- [ ] tidal.io.noaa

### CODE REVIEW

**Comment by ssolson** [comment]
@akeeste could you take a look at fixing the PR failed tests before I review?

```
_______________________ TestIO.test_load_usgs_data_daily _______________________

self = <mhkit.tests.river.test_io_usgs.TestIO testMethod=test_load_usgs_data_daily>

    def test_load_usgs_data_daily(self):
        file_name = join(datadir, "USGS_0831[30](https://github.com/MHKiT-Software/MHKiT-Python/actions/runs/8147648568/job/22268844047?pr=285#step:6:31)00_Jan2019_daily.json")
        data = river.io.usgs.read_usgs_file(file_name)
    
        expected_index = pd.date_range("2019-01-01", "2019-01-[31](https://github.com/MHKiT-Software/MHKiT-Python/actions/runs/8147648568/job/22268844047?pr=285#step:6:32)", freq="D")
        self.assertEqual(data.columns, ["Discharge, cubic feet per second"])
>       self.assertEqual((data.index == expected_index.tz_localize("UTC")).all(), True)
E       AssertionError: False != True
```

**Comment by akeeste** [comment]
> @akeeste could you take a look at fixing the PR failed tests before I review?
> 
> ```
> _______________________ TestIO.test_load_usgs_data_daily _______________________
> 
> self = <mhkit.tests.river.test_io_usgs.TestIO testMethod=test_load_usgs_data_daily>
> 
>     def test_load_usgs_data_daily(self):
>         file_name = join(datadir, "USGS_0831[30](https://github.com/MHKiT-Software/MHKiT-Python/actions/runs/8147648568/job/22268844047?pr=285#step:6:31)00_Jan2019_daily.json")
>         data = river.io.usgs.read_usgs_file(file_name)
>     
>         expected_index = pd.date_range("2019-01-01", "2019-01-[31](https://github.com/MHKiT-Software/MHKiT-Python/actions/runs/8147648568/job/22268844047?pr=285#step:6:32)", freq="D")
>         self.assertEqual(data.columns, ["Discharge, cubic feet per second"])
> >       self.assertEqual((data.index == expected_index.tz_localize("UTC")).all(), True)
> E       AssertionError: False != True
> ```

@ssolson yeah definitely, i'll see what's going on and ping you when it's ready

**Comment by akeeste** [comment]
@ssolson Sorry for the wait, I think I have the tests passing now. I'll double check tomorrow morning. I was having some issues converting ``river.io.usgs.py`` and especially ``variable_interpolation`` and ``turbulence_intensity`` in ``river.io.d3d.py`` to xarray. The calls to scipy.interp.griddata were proving difficult with d3d data in xr.Dataset format. For now, I reverted those functions to use pandas, but added the option to return xarray datasets. So river.io and tidal.io functions are not yet based in xarray, but can now return xarray.

**Comment by akeeste** [comment]
@ssolson okay tests are finally passing! I had to go back and fix a few changes I made to how some river.resource functions output their data, but that is all working as it was before. Again noting that the river and tidal io functions can now output xarray, but are not based in xarray yet because they were tricky

**Review by ssolson — COMMENTED** [review]
@akeeste thank you for putting this together. I only have a couple of comments. 

I see that the overlap in our util functions do not explicitly overlap. I think the name of my utils file is preferable e.g. I think `type_handling` is more descriptive of what these functions do than than `data_utils`.

The function I propose  in #288 ensures the data is numeric in addition to the data structure check. E.g. this function will currently accept ['a', 'b', 'c'] and then fail in calculation when performing operations on a string. 

Overall though I am not too worried about this and I am happy to mesh these functions together as I pass through with the linting in order to get our release out.

File: mhkit/tests/utils/test_utils.py [hunk 1]
```diff
@@ -159,6 +160,132 @@ def test_magnitude_phase_3D(self):
         self.assertTrue(all(theta == phase1))
         self.assertTrue(all(phi == phase2))
 
+    def convert_to_dataArray(self):
```
[the remark this answers is not in the document]
Comment (ssolson) [reply mhkit/tests/utils/test_utils.py:163]: ... bc this is chimera case (camel_snakeCase). xarray uses `dataarray` internally

File: mhkit/tidal/performance.py [hunk 2]
```diff
@@ -566,56 +558,13 @@ def device_efficiency(
...
-    timeseries is returned.
-
-    Parameters
-    -------------
-    power : xarray.DataArray or pandas.Series
-        The device power output timeseries.
-    U : xarray.DataArray
-        2D streamwise sea water velocity or sea water speed.
+    out = xr.Dataset({"U_avg": vel_hub, "Efficiency": eta})
+    out = out.rename({"speed_bins": "U_bins"})
```
[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/tidal/performance.py:562]: Agreed. I'll update the ``out`` variables with more descriptive names here, in ``power_curve`` and in ``_average_velocity_bins``

File: mhkit/utils/data_utils.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    if isinstance(data, xr.DataArray):
+        # xr.DataArray.to_dataset() breaks if the data variable is unnamed
+        if data.name == None:
+            data.name = name
+        data = data.to_dataset()
+
+    return data
+
+
+def convert_to_dataArray(data, name="data"):
```
[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/utils/data_utils.py:69]: Fair point, I'll update all these calls to ``convert_to_dataarray``

File: mhkit/tidal/performance.py [hunk 4]
```diff
@@ -437,14 +417,16 @@ def velocity_profiles(
         Time averaging window in seconds. Defaults to 600.
     func : string
         Function to apply. One of 'mean','rms', or 'std'
+    to_pandas: bool, optional
+        Flag to output pandas instead of xarray. Default = True.
 
     Returns
     ---------
     pandas.DataFrame
         Average velocity profiles based on ensemble mean velocity.
```
[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/tidal/performance.py:426]: Agreed, done.

File: mhkit/tests/tidal/test_performance.py [hunk 5]
```diff
@@ -79,6 +77,58 @@ def test_power_curve(
         assert_allclose(df93_circ.values[-2], test_circ, atol=1e-5)
         assert_allclose(df93_rect.values[-3], test_rect, atol=1e-5)
 
+    def test_power_curve_xarray(self):
+        df93_circ = performance.power_curve(
```
[the remark this answers is not in the document]
Comment (akeeste) [reply mhkit/tests/tidal/test_performance.py:81]: yeah this naming threw me off a bit, but I just went with what was there. For another PR...

**Comment by akeeste** [comment]
Thanks for the review @ssolson . I addressed all of your review comments, ported my functions to your much better named ``utils.type_handling`` 😊, and can confirm that the river, tidal, and tidal performance examples all run without errors or warnings. Tests are still passing

**Comment by akeeste** [comment]
Within test_utils, the name change from ``convert_to_dataarray`` to ``test_convert_to_dataarray`` made that test actually run. In ``convert_to_dataarray`` there was one minor inconsistency with how datasets are cast to dataarrays. the xr ``to_array`` function always makes the variable name a new dimension, which is not necessary in our scenario and makes the returned dataarray different from how all other types are converted. I fixed this case so the function works consistently

**Pull request merged**